contracts/interchain-token-service/InterchainTokenService.sol

```diff
@@ -533,28 +576,68 @@ contract InterchainTokenService is
-     * @param sourceAddress The address where the transaction originates from
+     * @param sourceService The address of the remote ITS where the transaction originates from
      * @param payload The encoded data payload for the transaction
      */
     function execute(
         bytes32 commandId,
         string calldata sourceChain,
-        string calldata sourceAddress,
+        string calldata sourceService,
         bytes calldata payload
-    ) external onlyRemoteService(sourceChain, sourceAddress) notPaused {
+    ) external onlyRemoteService(sourceChain, sourceService) notPaused {
         bytes32 payloadHash = keccak256(payload);
 
-        if (!gateway.validateContractCall(commandId, sourceChain, sourceAddress, payloadHash)) revert NotApprovedByGateway();
+        if (!gateway.validateContractCall(commandId, sourceChain, sourceService, payloadHash)) revert NotApprovedByGateway();
 
         uint256 selector = abi.decode(payload, (uint256));
-        if (selector == SELECTOR_RECEIVE_TOKEN || selector == SELECTOR_RECEIVE_TOKEN_WITH_DATA)
-            return _processReceiveTokenPayload(commandId, sourceChain, payload, selector);
+        if (selector == SELECTOR_RECEIVE_TOKEN || selector == SELECTOR_RECEIVE_TOKEN_WITH_DATA) {
+            address expressCaller = _popExpressCaller(commandId, payloadHash);
+            _processReceiveTokenPayload(expressCaller, sourceChain, payload, selector);
+
+            if (expressCaller != address(0))
+                emit ExpressExecutionFulfilled(commandId, sourceChain, sourceService, payloadHash, expressCaller);
+
+            return;
+        }
+
         if (selector == SELECTOR_DEPLOY_TOKEN_MANAGER) return _processDeployTokenManagerPayload(payload);
         if (selector == SELECTOR_DEPLOY_AND_REGISTER_STANDARDIZED_TOKEN) return _processDeployStandardizedTokenAndManagerPayload(payload);
 
         revert SelectorUnknown();
     }
 
+    function contractCallWithTokenValue(
+        string calldata /*sourceChain*/,
+        string calldata /*sourceAddress*/,
+        bytes calldata /*payload*/,
+        string calldata /*symbol*/,
+        uint256 /*amount*/
+    ) public view virtual returns (address, uint256) {
+        revert ExecuteWithTokenNotSupported();
+    }
+
+    function getExpressExecutorWithToken(
+        bytes32 /*commandId*/,
+        string calldata /*sourceChain*/,
+        string calldata /*sourceAddress*/,
+        bytes32 /*payloadHash*/,
+        string calldata /*symbol*/,
+        uint256 /*amount*/
+    ) external pure returns (address) {
+        revert ExecuteWithTokenNotSupported();
+    }
+
+    function expressExecuteWithToken(
+        bytes32 /*commandId*/,
+        string calldata /*sourceChain*/,
+        string calldata /*sourceAddress*/,
+        bytes calldata /*payload*/,
+        string calldata /*tokenSymbol*/,
+        uint256 /*amount*/
+    ) external payable {
+        revert ExecuteWithTokenNotSupported();
+    }
+
```

Comment:
We could remove these if we get contract size issues and they will still revert without an error code, but we can keep them in if we need to for clarity until then.

Reply:
We need them to satisfy the Express interface from GMP SDK